### PR TITLE
allow fetching REX keys from the API

### PR DESCRIPTION
**Pull request by evgeni**
this requires REX 1.6.7 or newer to be installed

includes #281 

so far untested :)

### CODE REVIEW

**Comment by fcami** [comment]
Not tested either, but that's the right way to do it. LGTM!

File: bootstrap.py [hunk 1]
```diff
@@ -1383,4 +1410,15 @@ def exec_service(service, command, failonerror=True):
             remove_obsolete_packages()
 
         if options.remote_exec:
-            install_foreman_ssh_key()
+            if options.remote_exec_proxies:
+                listproxies = options.remote_exec_proxies.split(",")
+                for proxy_fqdn in listproxies:
+                    remote_exec_url = "https://" + str(proxy_fqdn) + ":9090/ssh/pubkey"
```
Comment (ehelms) [inline bootstrap.py:1519]: Not great if we ever move around the smart proxy port or someone configures it differently. Fine to keep, worth noting.

File: bootstrap.py [hunk 2]
```diff
@@ -587,14 +580,44 @@ def install_foreman_ssh_key():
...
-        if foreman_ssh_key in open(foreman_ssh_authfile, 'r').read():
-            print_generic("Foreman's SSH key is already present in %s" % foreman_ssh_authfile)
+    install_ssh_key_from_string(foreman_ssh_key)
+
+
+def install_ssh_key_from_api():
+    """
+    Download and install all Foreman's SSH public keys.
+    """
+    url = "https://" + options.foreman_fqdn + ":" + str(API_PORT) + "/api/v2/smart_proxies/"
```
Comment (ehelms) [inline bootstrap.py:672]: This could use some apypie love (if it wasn't designed as a stand-alone script to install).

**Review by ehelms — APPROVED** [review]

File: bootstrap.py (hunk 1 above)
Comment (evgeni) [reply bootstrap.py:1519]: true, will add to the longish "refactor&improve" list :)

File: bootstrap.py (hunk 2 above)
Comment (evgeni) [reply bootstrap.py:672]: bootstrap.py has to run on python with only stdlib available :/

**Review by sideangleside — APPROVED** [review]
A couple of questions:

- if this requires a specific version of REX, is there any way to check for that version and fail gracefully if not? 
- Doesn't the API require auth for most things? so wouldn't you have to ensure that the user has authenticated when using this capability?

**Comment by evgeni** [comment]
> if this requires a specific version of REX, is there any way to check for that version and fail gracefully if not?

I can look into this. But the code will fail gracefully anyways, as in older versions there would be just no key listed, and then we don't add it. But we also don't tell the user that we didn't.

> Doesn't the API require auth for most things? so wouldn't you have to ensure that the user has authenticated when using this capability?

Correct, this only works if you pass user/password. Want me to add a check for that?

**Comment by ehelms** [comment]
Needs an ole rebase!

**Comment by evgeni** [comment]
rebased

**Pull request merged**